File: Python/BOJ/2805 나무 자르기.py
def search(arr):  # Parametric Search
    start = 1  # 1부터
    end = max(x)  # 최대치까지
    while start <= end:
        middle = (start + end) // 2  # 중간값을 기준으로
        num = 0  # 나무 갯수 0부터
        for i in x:
            if i-middle > 0:  # 높이에서 중간을 뺀게 양수면
                num += (i-middle)  # 나무 갯수 저장
        if num >= M:  # 목표치보다 크면
            start = middle + 1  # +1
        else:
            end = middle - 1
    return end  # 최종 높이값 반환


N, M = map(int, input().split())
x = list(map(int, input().split()))
print(search(x))

File: Java/BOJ/2805 나무자르기.java
import java.io.*;
import java.util.*;
import java.math.*;

public class Main {
	static long[] tree;
	static int N;
	static long M;
	static long search() {
		long start = 1;
		long end = tree[N-1];
		while (start <= end) {
			long mid = (start + end)/2;
			long num = 0;
			for (int i = 0; i< N; i++) {
				if (tree[i] > mid) {
					num += tree[i] - mid;					
				}
			}
			if (num >= M) {
				start = mid + 1;
			}
			else {
				end = mid - 1;
			}
		}
		return end;
	}
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        N = Integer.parseInt(st.nextToken());
        M = Long.parseLong(st.nextToken());
        st = new StringTokenizer(br.readLine(), " ");
        tree = new long[N];
        for(int i = 0; i < N; i++) {
        	tree[i] = Long.parseLong(st.nextToken());
        }
        Arrays.sort(tree);
        System.out.println(search());
    }
}